### PR TITLE
🚨 🚨  Raise error when no speaker embeddings in speecht5._generate_speech

### LINKED ISSUE #26401: [🚨 Bug] Text-to-Audio pipeline is broken for Speech T5 TTS
### System Info

- `transformers` version: 4.33.2
- Platform: Linux-5.15.120+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.17.2
- Safetensors version: 0.3.3
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.0.1+cu118 (False)
- Tensorflow version (GPU?): 2.13.0 (False)
- Flax version (CPU?/GPU?/TPU?): 0.7.2 (cpu)
- Jax version: 0.4.14
- JaxLib version: 0.4.14
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@sanchit-gandhi @ylacombe

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Speech T5 TTS expects speaker embeddings to be passed for understandable speech to be produced. At the moment in the `Text-to-Audio` pipeline we just call the model (processor) + generate without any speaker embeddings. This results in garbage audio as an output.

Full-repro here: https://github.com/Vaibhavs10/scratchpad/blob/main/transformers_speecht5_tts_repro.ipynb

### Expected behavior

Text-to-Audio pipeline for SpeechT5 TTS checkpoints outputs audio that is understandable.

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?

Following the discussion in #26401, this adds a warning when using `_generate_speech` without speaker embeddings. 
If necessary, and after discussion, we could throw an error instead of a warning!

Fixes #26401

## Who can review?

cc @Vaibhavs10 @sanchit-gandhi and @ArthurZucker

### CODE REVIEW

File: src/transformers/models/speecht5/modeling_speecht5.py [hunk 1]
```diff
@@ -2541,6 +2541,15 @@ def _generate_speech(
     vocoder: Optional[nn.Module] = None,
     output_cross_attentions: bool = False,
 ) -> Union[torch.FloatTensor, Tuple[torch.FloatTensor, torch.FloatTensor]]:
+    if speaker_embeddings is None:
+        logger.warning_once(
```
Comment (Vaibhavs10) [inline src/transformers/models/speecht5/modeling_speecht5.py:2545]: Shouldn't this be an exception? A warning often gets un-noticed by the end-user.
In this case the user would just not be able to generate anything without a speaker_embedding.

IMO we should make it pretty clear with an `Exception`.

File: src/transformers/models/speecht5/modeling_speecht5.py [hunk 2]
```diff
@@ -2541,6 +2541,15 @@ def _generate_speech(
...
+            """`speaker_embeddings` is not specified, which will probably generate unintelligible speech.
+                    Please, specify a `speaker_embeddings`. For example, you can use a speaker embeddings by following
+                    the code snippet provided in this link:
+                    https://huggingface.co/datasets/Matthijs/cmu-arctic-xvectors
```
Comment (Vaibhavs10) [inline src/transformers/models/speecht5/modeling_speecht5.py:2549]: Should we move this dataset to `microsoft` or `CMU` repo? Feels a bit weird having it under a personal namespace. 

Thoughts?

**Review by Vaibhavs10 — APPROVED** [review]
Thank you for the quick turn around on this! One discussion point and a nit.

IMO, this should be an exception over a warning.

File: src/transformers/models/speecht5/modeling_speecht5.py (hunk 1 above)
Comment (sanchit-gandhi) [reply src/transformers/models/speecht5/modeling_speecht5.py:2545]: Changing this code to require a speaker embedding would be a breaking change - it's possible that people are using fine-tuned versions of SpeechT5 without speaker embeddings, so throwing an error here would block them. I think a warning is sufficient

File: src/transformers/models/speecht5/modeling_speecht5.py (hunk 2 above)
Comment (sanchit-gandhi) [reply src/transformers/models/speecht5/modeling_speecht5.py:2549]: We'll have to ask Matthijs 😅

**Review by sanchit-gandhi — APPROVED** [review]
LGTM - thanks @ylacombe

**Comment by ylacombe** [comment]
Thanks for the reviews @sanchit-gandhi and @Vaibhavs10 !

I'll request @ArthurZucker and/or @LysandreJik's reviews on this tiny PR ! :hugs:

File: src/transformers/models/speecht5/modeling_speecht5.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/speecht5/modeling_speecht5.py:2545]: I think it comes down to whether or not we recommend this. I agree with @Vaibhavs10, we are trying to move away from warnings everywhere. And I also agree with @sanchit-gandhi an error is breaking. 
But given the issue, I think it makes sense to either raise an error, or use a default embedding?

**Review by ArthurZucker — COMMENTED** [review]
Would be more in favor of breaking and trying to be backward compatible as well. Thinking of a way to allow people who have finetuned with no speaker embedding: set the default to the value you would get in that case.

**Comment by ylacombe** [comment]
Hey @ArthurZucker, thanks for the review!

I'm not an expert on SpeechT5, but as @sanchit-gandhi said [here](https://github.com/huggingface/transformers/pull/26418#discussion_r1338953648):

> it's possible that people are using fine-tuned versions of SpeechT5 without speaker embeddings, so throwing an error here would block them.

Even more, even if it is fine-tuned with speaker embeddings, the default speaker embeddings for one version might be different for the default speaker embedding of another version.

In anyways, loading a default speaker embedding is done [here](https://huggingface.co/lysandre/text-to-speech-pipeline/blob/main/tts.py#L28-L30), with this snippet:


```python
from datasets import load_dataset

if speaker_embeddings is None:
  embeddings_dataset = load_dataset("Matthijs/cmu-arctic-xvectors", split="validation")
  speaker_embeddings = torch.tensor(embeddings_dataset[7305]["xvector"]).unsqueeze(0)
```

I'm not sure using `datasets` is compatible with `transformers`. How shoud we deal with this if we agree on a default speaker embedding ?

**Comment by Vaibhavs10** [comment]
The fine-tuned models also require passing a `speaker_embedding` btw: https://huggingface.co/learn/audio-course/chapter6/fine-tuning#speaker-embeddings

So the expected behaviour would always be to pass a `speaker_embedding` right? @sanchit-gandhi

**Comment by ArthurZucker** [comment]
You can register a non persistent buffer (only if the default is always the same)

**Comment by sanchit-gandhi** [comment]
Using a default embedding makes sense if we can do it in a non-breaking way, such as with the non-persistent buffer

**Comment by ylacombe** [comment]
Hey @sanchit-gandhi , @Vaibhavs10  and @ArthurZucker , feel free to quickly check the PR whenever you have time!

**Review by Vaibhavs10 — APPROVED** [review]
Good for me to merge! Thanks for working on this!

File: src/transformers/models/speecht5/modeling_speecht5.py (hunk 1 above)
Comment (sanchit-gandhi) [reply src/transformers/models/speecht5/modeling_speecht5.py]: Do you think we could instead register the default embedding as a [non-persistent buffer](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer) (i.e. a parameter that is not saved to the state dict) and initialise it as one of the default speaker embeddings?

Comment (ArthurZucker) [reply src/transformers/models/speecht5/modeling_speecht5.py]: I think it does yes!

**Review by ArthurZucker — APPROVED** [review]
Good for me. Let's add the breaking symbol 😉

**Comment by ylacombe** [comment]
Nice, merging!

**Pull request merged**